File: process/constructions/amsw_product.pl
#!/usr/bin/perl

my $name = 'amsw_product';
my $id;

$id = get_construction_id( $name, 5, 3 );

$constructions{ $name } = {};

$constructions{ $name }->{'process_array'} =
	sub {
		my $array = shift;

		return if ( $array->{'k'} > $MAXK );

		amsw_product_process_1( $array );

		if ( $array->{'t'} < $MAXT ) {
			amsw_product_process_2( $array );
		}

		amsw_product_process_3( $array );
	};

sub amsw_product_process_1 {
	my $array = shift;

	my $prev = get_max_k_array( 'k < ' . $array->{'k'} . ' AND t = ' . $array->{'t'} . ' AND v = ' . $array->{'v'} );
	my $min;
	$min = $prev ? ($prev->{'k'} + 1) : $array->{'v'};

	my $k = $array->{'k'};
	my $range = $k - $min;

	my $M0;
	my $M1;
	my $M2;

	my $arrays;
	if ( $array3in ) {
		$arrays = [ $array3in ];
	} else {
		$arrays = get_arrays( 't = ' . $array->{'t'} . ' AND v = ' . $array->{'v'} );
	}

	for ( $M1 = $array->{'t'} - 1; ($M1 <= $MAXV) && ($M1 < $k); $M1++ ) {
		if ( ($k % $M1) < $range ) {
			{
				use integer;
				$M0 = $k / $M1;
			}

			next if ($M0 <= 1);

			my $toobig = 0;

			foreach $array_3 ( @$arrays ) {
				my $M2 = $array_3->{'k'};

				my $array_2 = get_best_array( $M2, $M1, $array->{'t'} - 1 );
				next unless ( defined( $array_2 ) );

				my $newK = $M0 * $M2;
				if ( $newK >= $MAXK ) {
					$toobig = 1;
				}

				my $new_array = new_constructed_array( $array->{'N'} * $array_2->{'N'} + $array_3->{'N'},
				                            		   $newK,
													   $array->{'v'},
													   $array->{'t'} );

				$new_array->{'construction'} = $id;
				$new_array->{'ingredients'} = [ $array, $array_2, $array_3 ];
				$new_array->{'parameters'} = { M0 => $M0, M1 => $M1, M2 => $M2, base => ( $array3in ? 3 : 1 ) };

				if ( $PROCESS_MODE ) {
					print " -> "; print_array( $new_array ); print "\n";
				}

				persist_constructed_array( $new_array );

				last if $toobig;
			}
		}
	}
}

sub amsw_product_process_2 {
	my $array = shift;

	my $M2 = $array->{'k'};
	my $M1 = $array->{'v'};

	my $t = $array->{'t'} + 1;
	my $v;
	for ( $v = $t; $v <= $MAXV; $v++ ) {
		my $toobig = 0;

		my $array_3 = get_best_array( $M2, $v, $t );
		next unless ( defined( $array_3 ) );

		my $condition = "t = $t AND v = $v";
		my $arrays = get_arrays( $condition );

		my $M0;
		my $lastM0 = 0;

		my $array_1;
		foreach $array_1 ( @$arrays ) {
			{
				use integer;
				$M0 = $array_1->{'k'} / $M1;
			}

			next if ( $M0 == $lastM0 );
			$lastM0 = $M0;

			my $k = $M0 * $M2;
			if ( $k >= $MAXK ) {
				$toobig = 1;
			}

			my $new_array = new_constructed_array( $array_1->{'N'} * $array->{'N'} + $array_3->{'N'},
			                            		   $M0 * $M2,
												   $v,
												   $t );

			$new_array->{'construction'} = $id;
			$new_array->{'ingredients'} = [ $array_1, $array, $array_3 ];
			$new_array->{'parameters'} = { M0 => $M0, M1 => $M1, M2 => $M2, base => 2 };

			if ( $PROCESS_MODE ) {
				print " -> "; print_array( $new_array ); print "\n";
			}

			persist_constructed_array( $new_array );

			last if $toobig;
		}
	}
}

sub amsw_product_process_3 {
	my $array = shift;

	my $arrays = get_arrays( 't = ' . $array->{'t'} . ' AND v = ' . $array->{'v'} );
	my $array_1;

	foreach $array_1 ( @$arrays ) {
		amsw_product_process_1( $array_1, $array );
	}
}

1;

File: process/datastore/constructions/amsw_product.cpp
#include "construction.h"
#include "../datastore.h"
#include "../util/math.h"

#include <limits.h>
#include <iostream>

class c_amsw_product : public construction {
public:
	c_amsw_product() : construction( "amsw_product", 19 ) {
	}

	void process( array * base ) {
		if ( base->k > MAXK )
			return;

		process1( base );
		
		if ( base->t < MAXT ) {
			process2( base );
		}

		process3( base );
	}

	void process1( array * base, array * a3in = NULL, fullSortTableType * vulnerable = NULL ) {
		array * prev = getPreviousArray( base );
		int min = (prev != NULL) ? (prev->k + 1) : base->v;
		int k = base->k;
		int range = k - min;

		int M0, M1, M2;

		fullSortTableType arrays;
		if ( a3in ) {
			arrays.insert( a3in );
		} else {
			query( arrays, base->t, base->t, base->v, base->v, 0, -1 );
		}

		for ( M1 = base->t - 1; (M1 <= MAXV) && (M1 < k); M1++ ) {
			if ( (k % M1) < range ) {
                M0 = k / M1;

				if ( M0 <= 1 )
					continue;

				fullSortTableType::iterator it;
				for ( it = arrays.begin(); it != arrays.end(); it++ ) {
					array * array3 = *it;

					M2 = array3->k;
					array * array2 = getBestArray( M2, M1, base->t - 1 );

					if ( array2 == NULL )
						continue;

					__int64 lNewK = (__int64) M0 * (__int64) M2;
					if ( lNewK > INT_MAX  ) {
						//std::cout << "Avoided overflow of k." << std::endl;
						break;
					}
					int newK = (int) lNewK;

					array * arr = new array();

					__int64 lN = (__int64) base->N * (__int64) array2->N + (__int64) array3->N;
					if ( lN > INT_MAX  ) {
						//std::cout << "Avoided overflow of N." << std::endl;
						break;
					}
					arr->N = (int) lN;
					arr->k = newK;
					arr->v = base->v;
					arr->t = base->t;

					arr->type = 'C';
					arr->source = id;

					arr->ingredients.push_back( base );
					arr->ingredients.push_back( array2 );
					arr->ingredients.push_back( array3 );

					std::string value;
					int_to_string( M0, value );
					arr->parameters[ "M0" ] = value;

					int_to_string( M1, value );
					arr->parameters[ "M1" ] = value;

					int_to_string( M2, value );
					arr->parameters[ "M2" ] = value;

					int_to_string( (a3in != NULL ) ? 3 : 1, value );
					arr->parameters[ "base" ] = value;

					insertArray( arr, ( a3in != NULL ) ? vulnerable : &arrays );

					if ( newK >= MAXK )
						break;
				}
			}
		}
	}

	void process2( array * base ) {
		int M2 = base->k;
		int M1 = base->v;

		int t = base->t + 1;
		for ( int v = t; v <= MAXV; v++ ) {
			bool toobig = false;

			array * array3 = getBestArray( M2, v, t );
			if ( array3 == NULL )
				continue;

			fullSortTableType arrays;
			query( arrays, t, t, v, v, 0, -1 );

			int M0;
			int lastM0 = 0;

			fullSortTableType::iterator it;
			for ( it = arrays.begin(); it != arrays.end(); it++ ) {
				array * array1 = *it;

				M0 = array1->k / M1;

				if ( M0 == lastM0 )
					continue;

				lastM0 = M0;

				__int64 lNewK = (__int64) M0 * (__int64) M2;
				if ( lNewK > INT_MAX  ) {
					//std::cout << "Avoided overflow of k." << std::endl;
					break;
				}
				int k = (int) lNewK;

				array * arr = new array();

				__int64 lN = (__int64) array1->N * (__int64) base->N + (__int64) array3->N;
				if ( lN > INT_MAX  ) {
					//std::cout << "Avoided overflow of N." << std::endl;
					break;
				}
				arr->N = (int) lN;
				arr->k = k;
				arr->v = v;
				arr->t = t;

				arr->type = 'C';
				arr->source = id;

				arr->ingredients.push_back( array1 );
				arr->ingredients.push_back( base );
				arr->ingredients.push_back( cleanCopy( array3 ) );

				std::string value;
				int_to_string( M0, value );
				arr->parameters[ "M0" ] = value;

				int_to_string( M1, value );
				arr->parameters[ "M1" ] = value;

				int_to_string( M2, value );
				arr->parameters[ "M2" ] = value;

				int_to_string( 2, value );
				arr->parameters[ "base" ] = value;

				insertArray( arr, &arrays );

				if ( k >= MAXK )
					break;
			}

			deleteIfRule( array3 );
		}
	}

	void process3( array * base ) {
		fullSortTableType arrays;
		query( arrays, base->t, base->t, base->v, base->v, 0, -1 );

		fullSortTableType::iterator it;
		for ( it = arrays.begin(); it != arrays.end(); it++ ) {
			process1( *it, base, &arrays );
		}
	}
};

// declare the construction object
construction * amsw_product = new c_amsw_product();